### PR TITLE
Add tf.repeat support equivalent to numpy.repeat

### LINKED ISSUE #8264: Auto-Configuration Error: Cannot find cudnn.h at /usr/lib/x86_64-linux-gnu/include/cudnn.h
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
#4397, but it's for 16.04

### Environment info
Operating System: Ubuntu 14.04
Installed version of CUDA and cuDNN: 8.0, 5.1
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
```
ls -l /path/to/cuda/lib/libcud*
-rw-r--r-- 1 root 543K Jan 26 15:48 /usr/local/cuda/lib64/libcudadevrt.a
lrwxrwxrwx 1 root   16 Jan 26 15:51 /usr/local/cuda/lib64/libcudart.so -> libcudart.so.8.0
lrwxrwxrwx 1 root   19 Jan 26 15:51 /usr/local/cuda/lib64/libcudart.so.8.0 -> libcudart.so.8.0.61
-rw-r--r-- 1 root 406K Jan 26 15:48 /usr/local/cuda/lib64/libcudart.so.8.0.61
-rw-r--r-- 1 root 757K Jan 26 15:48 /usr/local/cuda/lib64/libcudart_static.a
```
```
ll /usr/lib/x86_64-linux-gnu/libcudnn*                                                                                                                                                                
lrwxrwxrwx 1 root  29 Mar  1 11:58 /usr/lib/x86_64-linux-gnu/libcudnn.so -> /etc/alternatives/libcudnn_so
lrwxrwxrwx 1 root  18 Nov  6 23:19 /usr/lib/x86_64-linux-gnu/libcudnn.so.5 -> libcudnn.so.5.1.10
-rw-r--r-- 1 root 81M Nov  6 23:19 /usr/lib/x86_64-linux-gnu/libcudnn.so.5.1.10
lrwxrwxrwx 1 root  32 Mar  1 11:58 /usr/lib/x86_64-linux-gnu/libcudnn_static.a -> /etc/alternatives/libcudnn_stlib
-rw-r--r-- 1 root 68M Nov  6 23:19 /usr/lib/x86_64-linux-gnu/libcudnn_static_v5.a
```

If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`): e895d5ca395c2362df4f5c8f08b68501b41f8a98
2. The output of `bazel version`: 
```
INFO: Reading 'startup' options from /root/.bazelrc: --batch
Build label: 0.4.2
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Dec 7 18:47:11 2016 (1481136431)
Build timestamp: 1481136431
Build timestamp as int: 1481136431
```

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)
I ran
```
./configure
```
It failed with
```
INFO: Reading 'startup' options from /root/.bazelrc: --batch
INFO: Starting clean (this may take a while). Consider using --expunge_async if the clean takes more than several minutes.
INFO: Reading 'startup' options from /root/.bazelrc: --batch
ERROR: package contains errors: tensorflow/compiler/tests.
ERROR: error loading package 'tensorflow/compiler/tests': Encountered error while reading extension file 'cuda/build_defs.bzl': no such package '@local_config_cuda//cuda': Traceback (most recent call last):
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 816
                _create_cuda_repository(repository_ctx)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 729, in _create_cuda_repository
                _get_cuda_config(repository_ctx)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 585, in _get_cuda_config
                _cudnn_version(repository_ctx, cudnn_install_base..., ...)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 296, in _cudnn_version
                _find_cuda_define(repository_ctx, cudnn_install_base..., ...)
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 271, in _find_cuda_define
                auto_configure_fail("Cannot find cudnn.h at %s" % st...))
        File "/tensorflow/third_party/gpus/cuda_configure.bzl", line 93, in auto_configure_fail
                fail("
%sAuto-Configuration Error:%s ...))

Auto-Configuration Error: Cannot find cudnn.h at /usr/lib/x86_64-linux-gnu/include/cudnn.h
```

I noticed the cudnn.h is in /usr/include/cudnn.h somehow, while its libs are in /usr/lib/x86_64-linux-gnu/.

I read cuda_configure.bzl, but didn't find an easy way to let it find both cudnn headers and libs.

### PR DESCRIPTION
**Pull request by yongtang**
(Most of the credit goes to RaggedTensor author, couldn't find the GitHub handle but the related commit is bad5d1a)

This PR tries to address the issue raised in #8246 to have tf.repeat equivalent to numpy.repeat.

Multiple attempts have been made in the past. The previous PR #15224 add the support with C++ which was closed, as a python implementation relying on existing ops is more desirable.

Now repeat has been added in ragged_util.py for RaggedTensor (the axis=None is not covered but this is fairly straight ward). The related commit is bad5d1a.

This PR
- moves the repeat implementation in ragged_util.py to array_ops.py so that it is possible to get exposed.
- adds additional logic to handle numpy.repeat with axis=None
- expose repeat_with_axis in ragged_util.py again so that existing logic in ragged tensor does not break
- expose `tf.repeat` in v1 and v2.

This fix fixes #8264.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by gunan** [comment]
I would not be the right reviewer for this.
You also need to have a TF API owner review this change, I think

**Comment by alextp** [comment]
@seanpmorgan I think this belongs in addons for now. WDYT?

**Comment by seanpmorgan** [comment]
Initial Thoughts:
1) Looks like a great contribution thanks @yongtang !
2) I'm struggling to find a location for this in addons. A key aspect that we're trying to push is addons must match an API pattern that we support (e.g. layers, optimizers, etc.) or it belongs to a sub-package that a reputable contributor or organization maintains (e.g. seq2seq). Is there a sub-package we could imagine this living under? It seems like a primative tensor transformation... so unless we make a tfa.transform or something it might be a little tough.

@karmel @facaiy thoughts?

**Comment by alextp** [comment]
@seanpmorgan should this be tfa.manip.repeat so it can conceivably end up in tf.manip.repeat?

**Comment by facaiy** [comment]
> Is there a sub-package we could imagine this living under? It seems like a primative tensor transformation... so unless we make a tfa.transform or something it might be a little tough.

I agree we need to create a new sub-package if addons accepts the contribution. But I'm not sure if it's appropriate to put primitive ops in addons, because the requirements in https://github.com/tensorflow/addons are:
>  contributions that conform to **well-established API patterns**, but implement new functionality not available in core TensorFlow.   

@karmel  @dynamicwebpaige  What do you think?

**Comment by seanpmorgan** [comment]
Yeah, my high level thought is that `tfa.manip` may not be an easily maintainable subpackage as it could be hard to define that expected API structure. Not impossible though and I'd be open to a subpackage submission (Currently working on the process of doing such a thing).

Apologies for the tepid answer, but obviously Addons will be in a forever battle with scope creep.

**Comment by karmel** [comment]
Maybe tf-text? I agree that it's not obvious that this fits inside the addons paradigm. Although, if it were an op for repeating with a python endpoint rather than python for the same, I guess that would be allowed.

**Comment by rthadur** [comment]
@yongtang gentle ping to check review comments.

**Comment by rmlarsen** [comment]
@yongtang could you please rebase?

**Comment by yongtang** [comment]
Sorry for so many people getting pinned for review. I think GitHub was confused by the rebase and push. I have remove unneeded reviews and pushed again.

**Comment by yongtang** [comment]
Thanks @rmlarsen. The PR has been rebased with merge conflict addressed. Please take a look.

**Comment by rmlarsen** [comment]
@alextp This seems like a natural companion to tf.tile, so why not accept the PR as is?

**Comment by alextp** [comment]
@rmlarsen no strong reason just trying to keep the API small and we couldn't determine quickly whether this was redundant. If you think it's useful I'll bring it up in the next API owners meeting to approve.

**Comment by rmlarsen** [comment]
@alextp I'm not 100% sure, so I'll defer to the tf-api owners. But since it doesn't involve any new kernels it seems like a handy utility compared to setting up, say, a convolution to achieve the same.

File: tensorflow/python/ops/array_ops.py
Comment (alextp) [inline tensorflow/python/ops/array_ops.py]: Can we add examples of inputs / outputs with different values of repeats and axis?

Does repeat need to be 1d?

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/array_ops.py
Comment (edloper) [reply tensorflow/python/ops/array_ops.py]: The documentation in repeat_with_axis is more detailed -- some of that detail should probably be copied here.  In particular:

* The requirement that `len(repeats)` must equal `data.shape[axis]` (or must equal `data.size` if axis is None).
* The fact that the size of the axis dimension in the returned value is `sum(repeats)`.
* The examples.

(And to answer alextp: repeats must be 1d or 0d.  If it's 0d, then it's broadcast to 1d.)

File: tensorflow/python/ops/array_ops.py
Comment (edloper) [inline tensorflow/python/ops/array_ops.py]: nit: 

    A Tensor which has the same shape as `input`, except along the given axis.
                                         ^^^^^^^

**Review by edloper — COMMENTED** [review]
Thanks for updating the docs!

**Comment by yongtang** [comment]
@edloper Thanks for the review. The PR has been updated. Please take a look.

**Review by alextp — DISMISSED** [review]

**Comment by yongtang** [comment]
@rthadur @alextp Previously, Kokoro CI test failed because of some API change in `test_util.TensorFlowTestCase`. 

I have fixed the build failure and all tests passes now, please take a look.

Sorry for not catching the test failure earlier.

**Review by alextp — APPROVED** [review]

**Comment by rthadur** [comment]
Merged internally , waiting for auto merge to happen.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by yongtang** [comment]
Looks like the PR has been merged into master so I will close this PR. Thanks everyone for the help to make this happen 👍 ❤️ 🎉 !

**Pull request closed without merging**
